connect/mirror/src/test/java/org/apache/kafka/connect/mirror/integration/MirrorConnectorsIntegrationExactlyOnceTest.java

```diff
@@ -46,4 +51,49 @@ public void startClusters() throws Exception {
         super.startClusters();
     }
 
+    @Override
+    @Test
+    public void testReplication() throws Exception {
+        super.testReplication();
+
+        // Augment the base replication test case with some extra testing of the offset management
+        // API introduced in KIP-875
+        // We do this only when exactly-once support is enabled in order to avoid having to worry about
+        // zombie tasks producing duplicate records and/or writing stale offsets to the offsets topic
+
+        String backupTopic1 = remoteTopicName("test-topic-1", PRIMARY_CLUSTER_ALIAS);
+        String backupTopic2 = remoteTopicName("test-topic-2", PRIMARY_CLUSTER_ALIAS);
+
+        // Explicitly move back to offset 0
+        // Note that the connector treats the offset as the last-consumed offset,
+        // so it will start reading the topic partition from offset 1 when it resumes
+        alterMirrorMakerSourceConnectorOffsets(backup, n -> 0L, "test-topic-1");
+        // Reset the offsets for test-topic-2
+        resetSomeMirrorMakerSourceConnectorOffsets(backup, "test-topic-2");
+        resumeMirrorMakerConnectors(backup, MirrorSourceConnector.class);
+
+        int expectedRecordsTopic1 = NUM_RECORDS_PRODUCED + ((NUM_RECORDS_PER_PARTITION - 1) * NUM_PARTITIONS);
+        assertEquals(expectedRecordsTopic1, backup.kafka().consume(expectedRecordsTopic1, RECORD_TRANSFER_DURATION_MS, backupTopic1).count(),
+                "Records were not re-replicated to backup cluster after altering offsets.");
+        int expectedRecordsTopic2 = NUM_RECORDS_PER_PARTITION * 2;
+        assertEquals(expectedRecordsTopic2, backup.kafka().consume(expectedRecordsTopic2, RECORD_TRANSFER_DURATION_MS, backupTopic2).count(),
+                "New topic was not re-replicated to backup cluster after altering offsets.");
+
+        @SuppressWarnings({"unchecked", "rawtypes"})
+        Class<? extends Connector>[] connectorsToReset = CONNECTOR_LIST.toArray(new Class[0]);
+        // Resetting the offsets for the heartbeat and checkpoint connectors doesn't have any effect
+        // on their behavior, but users may want to wipe offsets from them to prevent the offsets topic
+        // from growing infinitely. So, we include them in the list of connectors to reset as a sanity check
```

Comment:
Is this also the reason for not just throwing an `UnsupportedOperationException` in their `alterOffsets` methods? If so, we don't really need validation logic there right?

Reply:
Heh, I was wondering if we'd have to do a deep dive on the rationale here.

There are a few thoughts I had about what kind of offset tweaking to allow for the checkpoint and heartbeat connectors:
- (As noted in the comment here) Users may want to tombstone offsets for these connectors to prevent the offsets topic from growing infinitely
- We have no way of knowing if the set of offsets given to us is the total set or not, so we can't choose to only allow total resets instead of partial resets
- Since partial resets become possible, it seems reasonable to allow users to "undo" a partial reset by re-submitting an offset for a given partition
- This is especially relevant since, although the offsets topic isn't public API, its contents are now public API (via the `GET /connectors/{name}/offsets` endpoint), and users may want to track the offsets for these connectors for monitoring purposes
- We may make use of these offsets in future versions of these connectors, at which point, it'll be nice to know that the only offsets that we can ever encounter are the ones that were either emitted by the connector, or at least match the format of ones emitted by the connector

Also, I think the `UnsupportedOperationException` case was more intended for when offsets are managed externally, not for when it doesn't seem to make sense to modify them (because, e.g., they're never read back by the connector or its tasks).

So with all this in mind, I figured it'd be best to allow offsets to be modified by users, but only if they match the format of the offsets that the connector's tasks already emit. Thoughts?

As an aside... I'm wondering if we should abandon all kinds of validation when the proposed partition/offset pair is a tombstone. Just in case the offsets topic somehow ends up with garbage in it, it'd be nice to allow users to clean it up via the REST API, and if there's no existing partition/offset pair in the offsets topic for the given partition already, then emitting a tombstone effectively becomes a no-op. Thoughts x2?

Reply:
> (As noted in the comment here) Users may want to tombstone offsets for these connectors to prevent the offsets topic from growing infinitely

Since the set of source partitions is limited here, shouldn't log compaction be good enough?

> We have no way of knowing if the set of offsets given to us is the total set or not, so we can't choose to only allow total resets instead of partial resets

Yeah, that makes sense.

> Since partial resets become possible, it seems reasonable to allow users to "undo" a partial reset by re-submitting an offset for a given partition

If these offsets are never read back and actually used, why would users want to "undo" partial or complete resets?

> Also, I think the UnsupportedOperationException case was more intended for when offsets are managed externally, not for when it doesn't seem to make sense to modify them (because, e.g., they're never read back by the connector or its tasks).
>
> So with all this in mind, I figured it'd be best to allow offsets to be modified by users, but only if they match the format of the offsets that the connector's tasks already emit. Thoughts?

Hm, I still feel like if offset modification isn't really doing anything here it might be more intuitive to simply disallow it rather than informing users that the offsets have been modified successfully but with no actual side-effect. However, I don't think too many users are going to be attempting to modify the offsets for `MirrorCheckpointConnector` / `MirrorHeartbeatConnector` and expecting any change in their behavior so I'm not too fussed about this either way.

> As an aside... I'm wondering if we should abandon all kinds of validation when the proposed partition/offset pair is a tombstone. Just in case the offsets topic somehow ends up with garbage in it, it'd be nice to allow users to clean it up via the REST API, and if there's no existing partition/offset pair in the offsets topic for the given partition already, then emitting a tombstone effectively becomes a no-op

Hm this is an interesting one. I think if the offsets topic ends up with garbage in it through an external producer, users could always clean it up using the same external producer. If we don't let users write garbage using the offsets REST API, maybe we shouldn't let them clean up garbage using the REST API either? Overall, I think I'm more in favor of only allowing standard use cases with the offsets management REST APIs to reduce the potential number of footguns for users.

Reply:
> It's hard to define "limited" in this sense, but if you've configured MM2 to replicate every non-internal topic from a large cluster, then you could easily end up with hundreds of thousands of unique source offsets. And then, if some topics are deleted from the source cluster and other topics are created, we could go beyond even that. I think it'd be nice to allow people to do some cleanup in cases like that.

Thanks for the explanation, I think that makes sense.

> TL;DR: The contents of the offsets topic may become an additional point of observability for the connectors for people to discover, e.g., the total set of replicated topic partitions or consumer groups.

Ah, my bad, I did not clock that as a potential reason for the undo partial reset functionality. I agree that if we're deciding to allow tombstoning offsets for partitions, we should also allow the reverse operation.

> I definitely agree with the general mentality here of reducing the API surface in order to minimize the potential for users to hurt themselves. But is there any actual risk in the specific case of publishing tombstones for arbitrary topic partitions?

I don't think there's any real "risk" as such; more of a general concern on the flexibility - user friendliness tradeoff. Although thinking about this a bit more, a new / unfamiliar user can't really do any harm by accidentally writing a tombstone for an arbitrary partition (and an error response is also probably not gonna be terribly useful) so I think I'm okay with abandoning source partition validation when the source offset value is `null`. As an aside, my unfamiliarity with running or operating MM2 is making me averse to bikeshedding too much over this issue (FWICT these scenarios should be rare enough that any of the options we've discussed should be acceptable), so I'll defer to your expertise here 😄

Reply:
> I've pushed a change that allows arbitrary source partitions to be used with null source offsets; LMKWYT.

Thanks Chris, looks good. We can probably update the `FileStreamSourceConnector::alterOffsets` method with a similar change in a follow up as well.